common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,445 @@ public Collation(
+          };
+        }
+        return new Collation(
+          collationName(),
+          null,
+          comparator,
+          "1.0",
+          s -> (long) caseConversion(s).hashCode(),
+          collationId == UTF8_BINARY_COLLATION_ID,
+          collationId == UTF8_BINARY_COLLATION_ID,
+          collationId == UTF8_BINARY_LCASE_COLLATION_ID
+        );
+      }
+
+      private UTF8String caseConversion(UTF8String s) {
+        if (caseConversion == CaseConversion.LCASE) {
+          return s.toLowerCase();
+        } else if (caseConversion == CaseConversion.UCASE) {
+          return s.toUpperCase();
+        } else {
+          return s;
+        }
+      }
+
+      private String collationName() {
+        StringBuilder builder = new StringBuilder();
+        builder.append("UTF8_BINARY");
+        if (caseConversion != CaseConversion.UNSPECIFIED) {
+          builder.append('_');
+          builder.append(caseConversion.toString());
+        }
+        return builder.toString();
+      }
+    }
+
+    private static class CollationSpecICU extends CollationSpec {
+
+      private static final String[] ICULocaleNames;
+      private static final Map<String, ULocale> ICULocaleMap = new HashMap<>();
+      private static final Map<String, String> ICULocaleMapUppercase = new HashMap<>();
+      private static final Map<String, Integer> ICULocaleToId = new HashMap<>();
+      private static final String ICUCollatorVersion = "153.120.0.0";
+
+      static {
+        ICULocaleMap.put("UNICODE", ULocale.ROOT);
+        ULocale[] locales = Collator.getAvailableULocales();
+        for (ULocale locale : locales) {
+          if (locale.getVariant().isEmpty()) {
+            String language = locale.getLanguage();
+            assert (!language.isEmpty());
+            StringBuilder builder = new StringBuilder(language);
+            String script = locale.getScript();
+            if (!script.isEmpty()) {
+              builder.append('_');
+              builder.append(script);
+            }
+            String country = locale.getISO3Country();
+            if (!country.isEmpty()) {
+              builder.append('_');
+              builder.append(country);
+            }
+            String localeName = builder.toString();
+            assert (!ICULocaleMap.containsKey(localeName));
+            ICULocaleMap.put(localeName, locale);
+          }
+        }
+        for (String localeName : ICULocaleMap.keySet()) {
+          String localeUppercase = localeName.toUpperCase();
+          assert (!ICULocaleMapUppercase.containsKey(localeUppercase));
+          ICULocaleMapUppercase.put(localeUppercase, localeName);
+        }
+        ICULocaleNames = ICULocaleMap.keySet().toArray(new String[0]);
+        Arrays.sort(ICULocaleNames);
+        assert (ICULocaleNames.length <= (1 << 16));
+        for (int i = 0; i < ICULocaleNames.length; i++) {
+          ICULocaleToId.put(ICULocaleNames[i], i);
+        }
```

Comment:
The goal is to have stable mapping between integer ids and locale names. Please check generated golden [file](https://github.com/apache/spark/pull/46180/files#diff-23fa9a48d97f6a09df9ffbaffce69c040bbd1be2a6cf430242adf9597839b3e3) which outlines this mapping. It's also described as part of collation id [spec](https://github.com/apache/spark/pull/46180/files#diff-640c14aa5d7473df79b2435ce5a327dffcc16ca29354b153956b4f8d19fdb16cR133). Any changes to the mapping locale name - id mapping will require golden file update to add more visibility.

However, since the collation ids and therefore locale ids (as part of collation id) are now internal-only and never exposed to user (please confirm @dbatomic @stefankandic), they don't have to strictly remain the same across different Spark versions. That said, it could be problematic to have executors with different Spark versions with different locale name - id mapping.

Reply:
Overall design of collation ids in the PR is that `UTF8_BINARY` always has collation id 0. `UTF8_BINARY_LCASE` doesn't have collation id 1 anymore.

Reply:
Previous comment is outdated.

Resolving this conversation, clarified locale ids with comments around static members used in this static constructor and documented locale name - id mapping with golden file as well. New goal of `UTF8_BINARY` having collation id 0 and `UTF8_BINARY_LCASE` having collation id 1 is met with new changes.